packages/useful-types/test-d/types.test-d.ts

```diff
@@ -13,377 +8,288 @@ import type {
-
-interface Base {
-  required: string;
-  optional?: string;
-  nested?: Base;
-}
-
-expectAssignable<DeepPartial<Base>>({});
-expectAssignable<DeepPartial<Base>>({optional: 'test'});
-expectAssignable<DeepPartial<Base>>({required: undefined});
-expectAssignable<DeepPartial<Base>>({
-  nested: {nested: {nested: {required: 'test'}}},
-});
-
-interface ListObj {
-  list: Base[];
-}
-
-expectAssignable<DeepPartial<ListObj>>({list: []});
-expectAssignable<DeepPartial<ListObj>>({list: undefined});
-expectAssignable<DeepPartial<ListObj>>({
-  list: [{nested: {required: undefined}}],
+describe('ArrayElement', () => {
+  it('infers the array element type', () => {
+    interface Person {
+      firstName: string;
+    }
+
+    expect<ArrayElement<Person[]>>().type.toEqual<{firstName: string}>();
+    expect<ArrayElement<string[]>>().type.toBeString();
+    expect<ArrayElement<any[]>>().type.toBeAny();
+    expect<ArrayElement<(string | boolean)[]>>().type.toEqual<
+      string | boolean
+    >();
+  });
+
+  it('when `T` is not an array, resolves to the `never` type', () => {
+    expect<ArrayElement<string>>().type.toBeNever();
+  });
 });
 
-interface ReadOnlyListObj {
-  list: ReadonlyArray<Base>;
-}
-
-expectAssignable<DeepPartial<ReadOnlyListObj>>({list: []});
-expectAssignable<DeepPartial<ReadOnlyListObj>>({list: undefined});
-
-expectAssignable<DeepPartial<ReadOnlyListObj>>({
-  list: [{}],
-});
-
-expectNotAssignable<DeepPartial<ReadOnlyListObj>>({
-  list: [undefined],
-});
-
-expectAssignable<DeepPartial<string>>('test');
-
-interface RespectArrayElements {
-  list: (string | number)[] | null;
-}
-
-/**
- * valid types ar null or any array with string, number, or both
- */
-expectAssignable<DeepPartial<RespectArrayElements>>({list: null});
-expectAssignable<DeepPartial<RespectArrayElements>>({list: []});
-expectAssignable<DeepPartial<RespectArrayElements>>({list: ['string']});
-expectAssignable<DeepPartial<RespectArrayElements>>({list: [2]});
-expectAssignable<DeepPartial<RespectArrayElements>>({list: ['string', 2]});
-
-/**
- * handles ReadOnly objects
- */
-
-type PresetOption = Readonly<{
-  name?: string;
-  id: string;
-}>;
-
-type Preset = Readonly<{
-  options: ReadonlyArray<PresetOption>;
-  editableOptions: PresetOption[];
-}>;
-
-type PresetState = Readonly<{
-  list?: ReadonlyArray<Preset>;
-}>;
-type State = Readonly<{
-  preset: PresetState;
-}>;
-
-expectAssignable<DeepPartial<State>>({});
-expectAssignable<DeepPartial<State>>({
-  preset: {
-    list: [{options: [{name: 'string'}]}],
-  },
-});
-expectAssignable<DeepPartial<State>>({
-  preset: undefined,
+describe('DeepPartial', () => {
+  interface Base {
+    required: string;
+    optional?: string;
+    nested?: Base;
+  }
+
+  it('marks properties as optional recursively', () => {
+    expect<DeepPartial<Base>>().type.toMatch<{}>();
+    expect<DeepPartial<Base>>().type.toMatch<{optional?: string}>();
+    expect<DeepPartial<Base>>().type.toMatch<{required?: string}>();
```

Comment:
`.toMatch()` is used to check partial matching, otherwise `.toEqual()` is used. `.toEqual()` would not work here because the type is recursive.